ada_feeding/ada_feeding/behaviors/moveit2/moveit2_plan.py

```diff
@@ -0,0 +1,641 @@
+                                self.moveit2.set_path_orientation_constraint(
+                                    **constraint_kwargs
+                                )
+                            elif not self.blackboard_get(
+                                "ignore_violated_path_constraints"
+                            ):
+                                paths_satisfied = False
+                                break
+                    except tf2.LookupException:
+                        # Wait on Frame Transform
+                        # Use Timeout decorator to fail
+                        return py_trees.common.Status.RUNNING
+                    except (IndexError, AttributeError):
+                        self.logger.error(
+                            f"Malformed Goal Constraint: {constraint_kwargs}"
+                        )
+                        return py_trees.common.Status.FAILURE
+
+                # If any path constraints are unsatisfied, return FAILURE
+                if not paths_satisfied:
+                    self.moveit2.clear_goal_constraints()
+                    self.moveit2.clear_path_constraints()
+                    self.blackboard_set("trajectory", None)
+                    return py_trees.common.Status.FAILURE
+
+            ### Begin Planning
+            # pylint: disable=attribute-defined-outside-init
+            self.planning_future = self.moveit2.plan_async(
+                cartesian=self.blackboard_get("cartesian"),
+                max_step=self.blackboard_get("cartesian_max_step"),
+                start_joint_state=self.blackboard_get("start_joint_state"),
+            )
+            return py_trees.common.Status.RUNNING
+
+    @override
+    def terminate(self, new_status: py_trees.common.Status) -> None:
+        # Docstring copied from @override
+
+        # Clear constraints just in case
+        # Note plan_async should already do this
+        with self.moveit2_lock:
+            self.moveit2.clear_goal_constraints()
+            self.moveit2.clear_path_constraints()
+
+    def joint_constraint_satisfied(self, constraint_kwargs: Dict[str, Any]) -> bool:
+        """
+        Check if current joint state satisfies provided joint constraint.
+
+        Assume self.moveit2 is locked.
+
+        Parameters
+        ----------
+        constraint_kwargs: See moveit2.py; dictionary with keys: "joint_positions"
+                           "joint_names", "tolerance"
+        Returns
+        -------
+        True if the constraint is satisfied
+        """
+
+        # pylint: disable=protected-access
+        # We effectively need to to introspection
+        # to copy the functionality of MoveIt constraints.
+        # And this is read-only access.
+
+        curr_joint_state = self.moveit2.joint_state
+        tol = constraint_kwargs["tolerance"]
+
+        for i, des_position in enumerate(constraint_kwargs["joint_positions"]):
```

Comment:
des_position == desired_position

Reply:
I'm confused. I'm not making the assumption that constraint "joint_names" is the same as the publisher. that's what the call to `index` is for, and it happens every time the constraint joint_names is defined at the current index.

The problem with the proposed logic is that I don't think it handles the logic as-defined. If "joint_names"[i] exists, do the re-mapping. Otherwise, just borrow the order from the state publisher.

(In practice, for clarity, joint_names should be defined every time, but this code handles the case where it isn't).

Reply:
I can simplify the logic by requiring joint_names to be either len==0 (or None) or len(joint_positions)